text/0000-packages-as-optional-namespaces.md

```diff
@@ -0,0 +1,230 @@
+
+# Guide-level explanation
+
+The owners of the `foo` crate may provide other crates under the `foo` namespace, like `foo::bar`.  For users, this makes its official status clearer and makes it easier to discover.
+
+Users import these crates in Cargo.toml as normal:
+
+```toml
+[dependencies]
+"foo" = "1.0.42"
+"foo::bar" = "3.1"
+```
+
+They will then access this through a facade made of `foo` and all `foo::*` crates, for example:
+
+```rs
+let baz = foo::bar::Baz::new();
+foo::render(baz);
+```
+
+Some reasons for `foo`s owner to consider using namespaces:
+- Avoid name conflicts with third-party authors (since they are reserved)
+- Improve discoverability of official crates
+- As an alternative to feature flags for optional subsystems
+- When different parts of your API might have different compatibility guarantees
+
+When considering this, keep in mind:
+- Does it makes sense for this new crate to be presented in the `foo` facade?
+- How likely is a crate to move into or out of the namespace?
+  - Moving the crate in or out of a namespace is a breaking change though it can be worked around by having the old crate re-export the new crate but that does add extra friction to the process.
+  - There is not currently a mechanism to raise awareness with users that a crate has migrated into or out of a namespace and you might end up leaving users behind.
+- If users import both `foo` and `foo::bar` but `foo` also has a `bar` item in its API that isn't just `foo::bar` re-exported, then rustc will error.
+
+Only the owners of `foo` may _create_ the `foo::bar` crate (and all owners of `foo` are implicitly owners of `foo::bar`). After the `foo::bar` crate is created, additional per-crate publishers may be added who will be able to publish subsequent versions as usual.
+
+# Reference-level explanation
+
+`::` is now considered valid inside crate names on Crates.io. For now, we will restrict crate names to having a single `::` in them, not at the beginning or end of the name, but this can be changed in the future.
+
+When publishing a crate `foo::bar`, if the crate does not exist, the following must be true:
+
+ - `foo` must exist
+ - The user publishing the crate must be an owner of `foo`
+
+For the crate `foo::bar`, all owners of `foo` are always considered owners of `foo::bar`, however additional owners may be added. People removed from ownership of `foo` will also lose access to `foo::bar` unless they were explicitly added as owners to `foo::bar`.
+
+Crates.io displays `foo::bar` crates with the name `foo::bar`, though it may stylistically make the `foo` part link to the `foo` crate.
+
+The [registry index trie](https://doc.rust-lang.org/nightly/cargo/reference/registries.html#index-format) may represent subpackages by placing `foo::bar` as just `foo::bar`.
```

Comment:
Windows forbids `:` character in path names. The git index with a file named `foo::bar` will not be possible to check out on Windows.

Reply:
From a quick look, it appears the sensible options are `.+~@=^`. `[]{}` may also be options but are kinda ugly.

My top two choices would be `==` and `@@`.

Reply:
A single `@` seems cool; it is sufficiently clear about the intent (`foo@bar`)

Reply:
Isn't `@` already used to refer to `package@version`?

If `.` works that seems interesting as it has precedent with other packaging systems.

Reply:
I'm ultimately also fine with having the crates.io team pick an answer here; the registry trie is an internal implementation detail.

I think the experimental implementation uses `@`: https://github.com/integer32llc/crates.io-namespace-fork-index/tree/master/ic/u4/icu4x%40



> Isn't `@` already used to refer to `package@version`?

Not in the trie it isn't.

Reply:
I assume we are just talking about the layout of the index and not any user-facing syntax that can interfere with `package@version`

Reply:
Yep

Reply:
It seems like there's a general question of how to represent namespaced package names in a path/filename-specific way. This is also a problem for the naming of `.crate` files, and distribution packagers who need to represent namespaced crate names as distro-specific packages.

I think it would be a good idea for this RFC to stipulate filename-friendly encoding which can be consistently used for all of these use cases, rather than each use case inventing its own encoding.

Note that specifically for the distro package use case, the number of allowed non-alphanumberic characters is even smaller: for Debian, it's only `.+-` with the latter already allowed in crate names. That would leave `.` and `+` as the most widely compatible choices for a delimiter.

Reply:
As the index is not user facing, the design space is large.

I've not dug into cargo or crates.io's publish workflow to see how much we care about the file name of the package.

I suspect that overall the concerns and care abouts for all of these (including disributions) is disparate enough that they can inform each other but we shouldn't feel compelled for a single solution.

Reply:
The biggest issue is the index as that likely has the most forward/backwards compatibility concerns.

Beyond that, we need to consider how to not box ourselves in from using other characters in the future, like if we added support for traditional namespacing (I at least do not view this as replacing it).

For myself, I think the index and `.crate` file parts likely could be worked out between the cargo team and crates.io during implementation and stabilization.  I have these noted in the cargo's FCP issue for this RFC.

Reply:
For however much we do discuss distribution concerns, it'd be good if someone familiar with them could summarize how distribution packaging works for build packaging systems with native namespacing support.

Reply:
> For however much we do discuss distribution concerns, it'd be good if someone familiar with them could summarize how distribution packaging works for build packaging systems with native namespacing support.

Short summary: Pretty well, if well-defined.

One existing package ecosystem in Fedora that supports namespacing is Go (NPM packaging died in a :fire: a few years ago). And for Go packages, it is very awkward to map git repository URLs to package names (and not necessarily unique). For example, `github.com/prometheus/node_exporter` is currently mapped to `golang-github-prometheus-node-exporter` - and as you can already tell, that mapping is not unique. Most other language-specific package indices (that aren't dead) do *not* support namespacing and only have one flat namespace (PyPI, Rubygems, Haskell Hackage, Julia Packages, etc.).

The only ones I could find right now that use namespacing in their primary package index are Perl and PHP, and those have been doing that for a *long time* (and there's also no "non-namespaced" packages in those two ecosystems - it's always `Something::ModuleName` in Perl, or `vendor/name` in PHP, but there are *always* two identifiers, as far as I can tell).

So the way this proposed feature would work, with mixing both namespaced and non-namespaced names together, is really something that I don't think currently exists in other package ecosystems (other than NPM, and again, 🔥 ), and would need to be carefully considered for its implications.

Reply:
For Perl and PHP, what characters do you map `::` and `/` to?  And can you still run into the same issue that the character you used could end up in the package name or the namespace name and then cause conflicts?

For golang or any of the others, how important are unique mappings?

Reply:
In Perl and PHP, those are mapped to just a hyphen - which has not caused problems (mostly because in Perl, the "namespace" is more like a "category", and in PHP, it's the "vendor", so conflicts are very unlikely, as far as I can tell).

As for Go - it's not immediately clear how to map a repo URL to a `-`-separated name, so for example, these two projects would currently map to the same name (`golang-github-foo-bar-baz`):

- `github.com/foo-bar/baz`
- `github.com/foo/bar-baz`

And given the apparent predisposition of Go developers to fork projects constantly, this is not unlikely to cause name collisions at some point.

For other ecosystems (PyPI, Rubygems, etc.), where there is an "official" package registry, we attempt to ensure that our package names don't cause any collisions with packages that are in the official registry. For example, when packaging a Python package that's *not* published on PyPI, it's recommended to have the project at least *reserve* their name on PyPI - both to avoid name confusion attacks, and to avoid having to deal with name collisions if at some point a *different* project is published in PyPI with that name.

(The same reasons also hold for packages for Rust crates - we *exclusively* package "library" crates from crates.io, and no other sources are allowed.)

Reply:
> As for Go - it's not immediately clear how to map a repo URL to a --separated name, so for example, these two projects would currently map to the same name (golang-github-foo-bar-baz):

Go maps each segment in a [module path](https://go.dev/ref/mod#module-path) to an on-disk directory. For example, the module `github.com/ericlagergren/polyval` at commit `e25bc10ba391` is:

```
/Users/eric/go/pkg/mod/github.com/ericlagergren/polyval@v0.0.0-20220411101811-e25bc10ba391
```

Reply:
That doesn't work for mapping these names to package names for distros. Those can't contain slashes or colons.

Reply:
Yes, that's true. It does address the original concern, though:

> Windows forbids : character in path names. The git index with a file named foo::bar will not be possible to check out on Windows.

ISTM this RFC probably _shouldn't_ define a file name encoding.

Reply:
This RFC or the implementation _must_ define a file name encoding, both for how to lookup the crate in the index and how to map the resolved crate to a download URL, these are public APIs that have multiple implementations on both the client and server side <https://doc.rust-lang.org/nightly/cargo/reference/registry-index.html>

Reply:
To be clear, I see two separate things here:

- How to represent namespaces in the registry index.
- A general-purpose encoding format for package names. E.g.:

> [...] and distribution packagers who need to represent namespaced crate names as distro-specific packages.
>
> I think it would be a good idea for this RFC to stipulate a filename-friendly encoding for namespaced package names which can be consistently used for all of these use cases, rather than each use case inventing its own encoding.

Obviously this RFC must specify the former, but I'm not sure it needs to do the latter.

Reply:
Continuing this discussion in the crates.io tracing issue: https://github.com/rust-lang/crates.io/issues/8292#issuecomment-2000389870